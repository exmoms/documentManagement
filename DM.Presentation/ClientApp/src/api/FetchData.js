/**
 * @file this file includes functions that fetch data from API
 */

// Generic Delete Method
export async function deleteFromApi(url) {
  let res = await fetch(url, {
    method: "DELETE",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
    },
  });
  return res;
}

// Get MetaData Models
export async function fetchData(url) {
  let object = {};
  /*
   * send the request through a CORS proxy
   * https://stackoverflow.com/questions/43262121/trying-to-use-fetch-and-pass-in-mode-no-cors/43268098#43268098
   */
  await fetch(url, {
    method: "GET",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
    },
  })
    .then((response) => response.json())
    .then(
      (data) => (object = data),
      (error) => console.log(error)
    );
  return object;
}
