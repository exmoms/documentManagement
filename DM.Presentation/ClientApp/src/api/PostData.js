/**
 * @file This file includes functions that help to post data to a specific API
 * @version 1.0
 */

/**
 * Post new document data.
 * Upload form-data, scanned copy of doc and attachments.
 * @author Youssef Shaaban <y.shaaban@lit-co.net>
 * @param {*} documentData
 * @param {*} images
 * @param {*} attachments
 * @param {*} url
 * @param {*} option its values either "Add" for add new document version or "Update" to update exsiting document Version
 */
export async function postDocumentData(
  documentData,
  scannedPages,
  attachments,
  url = "",
  option
) {
  var formData = new FormData();

  if (process.env.NODE_ENV === "development") {
    console.log(documentData);
    console.log(attachments);
    console.log(scannedPages);
  }

  switch (option) {
    case "Add":
      formData.append("Document", JSON.stringify(documentData));
      break;
    case "Update":
      formData.append("Version", JSON.stringify(documentData));
      break;
    default:
  }

  attachments.forEach((attachment) => {
    formData.append("Attachments", attachment);
  });
  scannedPages.forEach((scannedPage) => {
    formData.append("Scans", scannedPage);
  });

  let response = {};
  await fetch(url, {
    // content-type header should not be specified!
    method: "POST",
    headers: {
      Accept: "application/json",
      // "Content-Type": "application/json",
    },
    body: formData,
  }).then(
    (res) => {
      response = res;
      // response.json();
    },
    (error) => console.log("[ERROR]:", error)
  );
  return response;
}

/**
 * Post Attachment Data.
 * Upload attachment json data + attachment file  .
 * @author Ali Daghman <ali.daghman@lit-co.net>
 * @param {*} payload
 * @param {*} attachmentFile
 * @param {*} url
 */
export async function postAttachmentData(payload, attachmentFile, url = "") {
  var formData = new FormData();

  formData.append("Attachment", JSON.stringify(payload));
  formData.append("FileAttachment", attachmentFile);

  let response = {};
  await fetch(url, {
    // content-type header should not be specified!
    method: "POST",
    body: formData,
  }).then(
    (res) => {
      response = res;
      // response.json();
    },
    (error) => console.log("[ERROR]:", error)
  );
  return response;
}

/**
 * Post Request To API
 * Send post request to api with  payload  and return a response.
 * @author Ali Daghman <ali.daghman@lit-co.net>
 * @param {*} url The api url
 * @param {*} payload js object which will be sent as request body.
 */
export async function postDataToAPI(url, payload = null, method = "POST") {

  const response = await fetch(url, {
    method: method,
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });
  return response;
}
