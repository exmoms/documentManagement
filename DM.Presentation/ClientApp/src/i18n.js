import i18n from "i18next";
import Backend from "i18next-xhr-backend";
import { initReactI18next } from "react-i18next";

var middleware = require("i18next-express-middleware");

const lngDetectorOptions = {
  // order and from where user language should be detected
  order: [/*"path", "session", "querystring",*/ "localStorage" /*"header"*/],

  // keys or params to lookup language from
  lookupLocalStorage: "i18nextLng",
};

i18n
  .use(initReactI18next)
  .use(middleware.LanguageDetector)
  .use(Backend)
  .init({
    preload: ["en", "ar"],
    initImmediate: false,
    whitelist: ["en", "ar"],
    lng:
      localStorage.getItem("i18nextLng") ||
      "en" /* don't uncomment this if you want to get lang from lang detector */,
    backend: {
      /* translation file path */
      loadPath: "/assets/i18n/{{ns}}/{{lng}}.json",
    },
    detection: lngDetectorOptions,
    load: "languageOnly",
    fallbackLng: "en",
    debug: false,
    /* can have multiple namespace, in case you want to divide a huge translation into smaller pieces and load them on demand */
    ns: ["translations"],
    defaultNS: "translations",
    keySeparator: false,
    interpolation: {
      escapeValue: false,
      formatSeparator: ",",
    },
    react: {
      wait: true,
    },
  });

export default i18n;
