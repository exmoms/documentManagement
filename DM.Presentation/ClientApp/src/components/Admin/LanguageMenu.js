import Button from "@material-ui/core/Button";
import Menu from "@material-ui/core/Menu";
import MenuItem from "@material-ui/core/MenuItem";
import TranslateIcon from "@material-ui/icons/Translate";
import React, { useState } from "react";
import { useCookies } from "react-cookie";
import "../../i18n";
import { useTranslation } from "react-i18next";
import { postDataToAPI } from "../../api/PostData";
import { Notification } from "../Admin/Notifications";

const langMap = { en: "English", ar: "العربية", de: "Deutsch" };

export default function LanguageMenu() {
  const [anchorEl, setAnchorEl] = React.useState(null);

  const { i18n } = useTranslation();

  const [cookies] = useCookies(["lang"]);
  var lang = cookies.lang ? cookies.lang.substring(2, 4) : "en";

  localStorage.setItem("i18nextLng", lang);

  const handleClick = (event) => {
    setAnchorEl(event.currentTarget);
  };

  const handleClose = () => {
    setAnchorEl(null);
  };

  const [open, setOpen] = useState(false);
  const [errorMessage, seterrorMessage] = useState([]);
  const [error, seterror] = useState(false);

  const setLanguage = async (event, langCode) => {
    lang = langCode;
    let response = await postDataToAPI("/api/User/ChangeLanguage?lang=" + lang);
    if (response.ok) {
      handleClose();
      i18n.changeLanguage(lang);
      localStorage.setItem("i18nextLng", lang);
    } else {
      // let json = await response.json();
      setOpen(true);
      seterror(true);
      seterrorMessage(["Unable to change language.."] /* json.error */);
    }
  };

  return (
    <React.Fragment>
      <Notification
      id="notification"
        open={open} // must be a "state" to show notification message [true or false]
        setOpen={setOpen} // this function changes the state "open" value.
        error={error} // if error message put [true] else for success put [false]
        errorMessage={errorMessage} // the recived error message (array of string) from server.
      />
      <Button
        aria-controls="language-menu"
        aria-haspopup="true"
        onClick={handleClick}
        variant="contained"
        color="primary"
        startIcon={<TranslateIcon />}
      >
        {langMap[lang]}
      </Button>
      <Menu
        id="language-menu"
        anchorEl={anchorEl}
        keepMounted
        open={Boolean(anchorEl)}
        onClose={handleClose}
      >
        <MenuItem onClick={(event) => setLanguage(event, "en")}>
          English
        </MenuItem>
        <MenuItem onClick={(event) => setLanguage(event, "ar")}>
          العربية
        </MenuItem>
      </Menu>
    </React.Fragment>
  );
}
