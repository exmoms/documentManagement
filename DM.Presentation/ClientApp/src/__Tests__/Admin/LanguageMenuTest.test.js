
import React from "react";
import ReactDom from "react-dom";
import LanguageMenu from "../../components/Admin/LanguageMenu";
import { cleanup } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import { shallow , mount} from "enzyme";
import toJSON from "enzyme-to-json";
import { MemoryRouter } from "react-router-dom";
import renderer from "react-test-renderer";
import { Link } from "react-router-dom";
import { Suspense } from "react";
import * as FetchMock from "../../api/FetchData";
import * as PostDataMock from "../../api/PostData";
import { useCookies } from "react-cookie";
import MenuItem from "@material-ui/core/MenuItem";
import { act } from "react-dom/test-utils";
import List from "@material-ui/core/List";
import Visibility from "@material-ui/icons/Visibility";
import VisibilityOff from "@material-ui/icons/VisibilityOff";
import {Redirect } from "react-router-dom";
import Typography from "@material-ui/core/Typography";
import LinearProgress from "@material-ui/core/LinearProgress";
import Button from "@material-ui/core/Button";
import i18n from "i18next";
afterEach(cleanup);
describe("LanguageMenu", () => {
    beforeEach(()=>{
        jest.clearAllMocks();
          i18n
    .init({
        react: {
            useSuspense: false
        }
    }); 
    jest.mock('react-cookie', () => ({
        useCookies: () => ({
          lang: "french"
        })
      }));
        jest.mock('react-i18next', () => ({
            useTranslation: () => ({
              i18n: { changeLanguage: jest.fn() }
            })
          }));
    })
    afterEach(() => {
    jest.clearAllMocks();
  });
  test("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDom.render(<LanguageMenu/>
        ,div
    );
  });

  test("matches snapshot", () => {
    const wrapper = shallow(<LanguageMenu/>);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
  
  test("intialization LanguageMenu", () => {
      const setItemMock = jest.spyOn(window.localStorage.__proto__, 'setItem');
      const wrapper = mount(<LanguageMenu/>);
      expect(setItemMock).toHaveBeenCalled();
      expect(setItemMock.mock.calls[0][0]).toBe("i18nextLng");
      expect(setItemMock.mock.calls[0][1]).toBe("en");
      expect(wrapper.find(LanguageMenu).find(Button).text()).toBe("English");
    expect(wrapper.find(LanguageMenu).find("#language-menu").at(0).props().anchorEl).toBe(null);
    expect(wrapper.find(LanguageMenu).find("#language-menu").at(0).props().open).toBeFalsy();
    setItemMock.mockClear();
  });

  test("handleClick ,handleClose ", () => {
    const wrapper = mount(<LanguageMenu/>);
    act(() => {
        wrapper
          .find(Button)
          .at(0)
          .props()
          .onClick({ currentTarget: { } });
      });
      wrapper.update();
      wrapper.render();
    expect(wrapper.find(LanguageMenu).find(Button).text()).toBe("English");
  expect(wrapper.find(LanguageMenu).find("#language-menu").at(0).props().anchorEl).toStrictEqual({});
  expect(wrapper.find(LanguageMenu).find("#language-menu").at(0).props().open).toBeTruthy();

  act(() => {
    wrapper
      .find("#language-menu")
      .at(0)
      .props()
      .onClose({ currentTarget: { } });
  });
  wrapper.update();
  wrapper.render();
  expect(wrapper.find(LanguageMenu).find("#language-menu").at(0).props().anchorEl).toBe(null);
  expect(wrapper.find(LanguageMenu).find("#language-menu").at(0).props().open).toBeFalsy();
});

test("setLanguage when res is 200", () => {
    const wrapper = mount(<LanguageMenu/>);
    const setItemMock2 = jest.spyOn(window.localStorage.__proto__, 'setItem');
const mockSuccessResponsePost ={ok :true};
const mockJsonPromisePost = Promise.resolve(mockSuccessResponsePost);
const PostMock = jest.spyOn(PostDataMock, "postDataToAPI").mockImplementation(() => mockJsonPromisePost);
act(() => {
    wrapper
      .find(Button)
      .at(0)
      .props()
      .onClick({ currentTarget: { } });
  });
  wrapper.render();
  wrapper.update();
  expect(wrapper.find(LanguageMenu).find("#language-menu").at(0).props().anchorEl).toStrictEqual({});
  expect(wrapper.find(LanguageMenu).find("#language-menu").at(0).props().open).toBeTruthy();
return mockJsonPromisePost.then(() => {
    act(() => {
        wrapper
          .find(MenuItem)
          .at(1)
          .props()
          .onClick({ target: {value:"" } });
      });
}).then(() => {
    wrapper.update();
    wrapper.render();
    expect(PostMock).toHaveBeenCalledTimes(1);
  expect(PostMock.mock.calls[0][0]).toBe("/api/User/ChangeLanguage?lang=ar");
  expect(wrapper.find(LanguageMenu).find(Button).text()).toBe("English");
  expect(setItemMock2).toHaveBeenCalled();
  expect(wrapper.find(LanguageMenu).find("#language-menu").at(0).props().anchorEl).toBe(null);
  expect(wrapper.find(LanguageMenu).find("#language-menu").at(0).props().open).toBeFalsy();
  setItemMock2.mockClear();
});

});

test("setLanguage when res is not 200", () => {
    const wrapper = mount(<LanguageMenu/>);
    const alertMock = jest.spyOn(window, 'alert');
const mockSuccessResponsePost ={ok :false};
const mockJsonPromisePost = Promise.resolve(mockSuccessResponsePost);
const PostMock = jest.spyOn(PostDataMock, "postDataToAPI").mockImplementation(() => mockJsonPromisePost);
return mockJsonPromisePost.then(() => {
    act(() => {
        wrapper
          .find(MenuItem)
          .at(1)
          .props()
          .onClick({ target: {value:"" } });
      });
}).then(() => {
    wrapper.update();
    wrapper.render();
    expect(PostMock).toHaveBeenCalledTimes(1);
  expect(PostMock.mock.calls[0][0]).toBe("/api/User/ChangeLanguage?lang=ar");
  expect(wrapper.find(LanguageMenu).find(Button).text()).toBe("English");
  expect(wrapper.find(LanguageMenu).find("#language-menu").at(0).props().anchorEl).toBe(null);
  expect(wrapper.find(LanguageMenu).find("#language-menu").at(0).props().open).toBeFalsy();
        expect(wrapper.find("#notification").props().open).toBeTruthy(); 
      expect(wrapper.find("#notification").props().error).toBeTruthy(); 
      expect(wrapper.find("#notification").props().errorMessage).toStrictEqual(["Unable to change language.."]); 
});

});
});