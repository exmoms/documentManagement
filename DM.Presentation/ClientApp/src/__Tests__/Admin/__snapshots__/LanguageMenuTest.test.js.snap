// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageMenu matches snapshot 1`] = `
<Fragment>
  <Notification
    error={false}
    errorMessage={Array []}
    id="notification"
    open={false}
    setOpen={[Function]}
  />
  <WithStyles(ForwardRef(Button))
    aria-controls="language-menu"
    aria-haspopup="true"
    color="primary"
    onClick={[Function]}
    startIcon={<UNDEFINED />}
    variant="contained"
  >
    English
  </WithStyles(ForwardRef(Button))>
  <WithStyles(ForwardRef(Menu))
    anchorEl={null}
    id="language-menu"
    keepMounted={true}
    onClose={[Function]}
    open={false}
  >
    <WithStyles(ForwardRef(MenuItem))
      onClick={[Function]}
    >
      English
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      onClick={[Function]}
    >
      العربية
    </WithStyles(ForwardRef(MenuItem))>
  </WithStyles(ForwardRef(Menu))>
</Fragment>
`;
